### PR TITLE
Additional log fields for unauthorized requests

### PR DESCRIPTION
**Pull request by mmazour**
### Description

Closes https://github.com/overleaf/write_latex/issues/3841

For improved forensics and customer support, add more fields to the `Unauthorized` lines in the log:
- Username
- IP address
- Status code returned by the OAuth attempt

#### Screenshots

Before
```
2017-08-09 13:45:29.859 INFO  [qtp988458918-19] GitBridgeApp: [245jzgbyzywspqc] Unauthorized
```

After (when auth failure)
```
2017-09-08 09:08:27.568 INFO  [qtp1629604310-23] GitBridgeApp: [245jzgbyzywspqc] Unauthorized, User 'michael.mazour@overleaf.com' status=401 ip=192.168.33.1
```

After (when no auth provided and we are going to signal Git to request credentials and then retry)
```
2017-09-08 09:00:49.230 INFO  [qtp1629604310-22] GitBridgeApp: [245jzgbyzywspqc] Unauthorized, User '(unknown)' status=0 ip=192.168.33.1
```

#### Related Issues / PRs

https://github.com/overleaf/write_latex/issues/3841

### Review

- The cases that can invoke `Unauthorized` logging are part of a large nested `if` with some `try/catch` clauses for good measure.  I consider this sort of code inherently brittle to changes, but fortunately I have not needed to tamper with the logic, only with the arguments to the three calls to the `unauthorized` method.

- To ensure the username was available at the point where `unauthorized` is called, I have added a variable at the top of the method. Although this feels inelegant, it let me avoid changes to the nested-if logic, which I felt was more important.

- I changed the `unauthorized` method signature to expect an `HttpServletResponse` instead of a `ServletResponse`, since the `response` passed in by the caller has already been cast to `HttpServletResponse` anyway.  I believe this should be fully safe (and IntelliJ didn't complain), but it's not inconceivable that I've overlooked some subtle reason Winston hadn't already done that. 

- To my surprise there really aren't tests devoted to this class.  Since this is only a logging change and does not alter the logic flow, I did not add any.

FWIW I used IntelliJ CE for these changes and its static code analysis is impressive (which is to say it's a harsh and relentless critic).  As a result do I have a pretty high comfort level with these changes (in addition to all the tests passing of course).

#### Potential Impact

- Can only affect projects that require authentication.
- _Should_ only affect log output.
- Potentially _could_ cause exceptions when `unauthorized` is called if I have mis-coded any cases (but I do not believe I have).
- Will slightly increase size of log files going forward.

#### Testing Checklist

- [x] Test in development environment (git-bridge VM + Overleaf VM)

### Deployment

- Normal deployment.
- [ ] Test no-auth-provided path and watch log
- [ ] Test wrong-auth-provided path and watch log
- [ ] Test correct-auth-provided path and watch log

#### Documentation

None.

#### Metrics and Monitoring

Nothing special.

#### Who Needs to Know?

- @jdleesmiller 
- @winstonli 
- @henryoswald 
- @Hugh-OBrien

### CODE REVIEW

File: src/main/java/uk/ac/ic/wlgitbridge/server/Oauth2Filter.java [hunk 1]
```diff
@@ -135,15 +139,15 @@ private void getAndInjectCredentials(
                                     servletResponse
                             );
                         } else {
-                            unauthorized(projectName, response);
+                            unauthorized(projectName, capturedUsername, 0, request, response);
```
Comment (winstonli) [inline src/main/java/uk/ac/ic/wlgitbridge/server/Oauth2Filter.java:142]: Perhaps we can use the 401 from the response we give in `unauthorized()`?

**Review by winstonli — COMMENTED** [review]
Sorry about the code, it's the worst in the git bridge by far. From reading over it, none of your changes break anything, and we assume `username` can never be null anyway.

**Review by jdleesmiller — APPROVED** [review]
Looks good!

**Comment by jdleesmiller** [comment]
Aha, I see our reviews crossed. Henry's at an event today, so I've assigned back to @mmazour .

**Review by winstonli — APPROVED** [review]

**Comment by winstonli** [comment]
Yep, anyway, looks good! Again sorry about the surrounding code...

**Comment by mmazour** [comment]
Thanks, Winston!  I'll look at that tweak you suggested.

File: src/main/java/uk/ac/ic/wlgitbridge/server/Oauth2Filter.java (hunk 1 above)
Comment (mmazour) [reply src/main/java/uk/ac/ic/wlgitbridge/server/Oauth2Filter.java:142]: @winstonli My intention here was to return the status code of the OAuth request to Overleaf, which I think I'm doing on line 127.  If I'm reading the code right, if we reach the `unauthorized` on line 142 then the OAuth request hasn't been made, so I'm arbitrarily showing `0` just to reflect that this is a different sort of `unauthorized`.

Comment (winstonli) [reply src/main/java/uk/ac/ic/wlgitbridge/server/Oauth2Filter.java:142]: Ok, that makes perfect sense.

Comment (mmazour) [reply src/main/java/uk/ac/ic/wlgitbridge/server/Oauth2Filter.java:142]: OK, thanks 👍

**Pull request merged**